README.md
# Scattering-of-Charged-Particles
A few iterations of the module.


changelog
1: The calculation of the paths of two different charged particles interacting with each other

2: The calculation of the path of an alpha particle interacting with a fixed gold nucleus

3: The calculation of the paths of a lot of alpha particles interacting with a fixed golden nucleus

4: Finding the relation of h and the angle

5: Finding the relation of the nucleus charge against the area effect (less than 1 degree) (linear - linear)

5: Finding the relation of the nucleus charge against the area effect (less than 1 degree) (log - log)

iteration 2.py
import matplotlib.pyplot as plt
import numpy as np

# moving particle and resting particle
class par: # particle class

  #vacuum electric permittivity
  vep = 8.8541878128 * 10**-12

# self, mass, electric charge, x acceleration, y acceleration, x velocity, y velocity, x position, y position
  def __init__(self, m, e, xa, ya, xv, yv, xp, yp): 
    self.m = m # (kg)
    self.e = e # (coulomb)
    self.xa = xa # (m/s^2)
    self.xas = [xa]
    self.ya = ya # (m/s^2)
    self.yas = [ya]
    self.xv = xv # (m/s)
    self.xvs = [xv]
    self.yv = yv # (m/s)
    self.yvs = [yv]
    self.xp = xp # (m)
    self.xps = [xp]
    self.yp = yp # (m)
    self.yps = [yp]
  
  def next_position(self):
    self.xp = self.xp + self.xv * time_interval
    self.xps.append(self.xp)
    self.yp = self.yp + self.yv * time_interval
    self.yps.append(self.yp)

  def next_velocity(self):
    self.xv = self.xv + self.xa * time_interval
    self.xvs.append(self.xv)
    self.yv = self.yv + self.ya * time_interval
    self.yvs.append(self.yv)
  
  def next_acceleration(self, other):
    dx = abs(self.xp - other.xp)
    dy = abs(self.yp - other.yp)
    self.xa = (dx * self.e * other.e) / ((dx**2 + dy**2)**1.5 * 4 * np.pi * par.vep * self.m)
    if self.xp < other.xp:
      self.xa = -self.xa
    self.xas.append(self.xa)
    self.ya = (dy * self.e * other.e) / ((dx**2 + dy**2)**1.5 * 4 * np.pi * par.vep * self.m)
    if self.yp < other.yp:
      self.ya = -self.ya
    self.yas.append(self.ya)

# main computational loop
time_interval = 0.2
total_time = 12
ts = np.arange(0, total_time+time_interval, time_interval)

# ap1 is an alpha particle
apm = 6.6446573357 * 10**-27
apc = 2 * 1.602176634 * 10**-19
ap1 = par(apm, apc, 0, 0, 2, 0, 0, 3)

# par2 is an gold nucleus
gnm = 196.966543 * 1.660539 * 10**-27
gnc = 79 * 1.602176634 * 10**-19
gn = par(gnm, gnc, 0, 0, 0, 0, 10, 0)

for t in ts[1:]:
  ap1.next_position()
  ap1.next_velocity()
  ap1.next_acceleration(gn)

print(ap1.e)
# plotting the path of both particles
plt.plot(ap1.xps, ap1.yps)
plt.plot(10,0,'ro') 
plt.xlabel('X')
plt.ylabel('Y')
plt.title('The path of both particles in 2D')
plt.show()

'''
# plotting 2d
fig, axs = plt.subplots(2,3)

axs[0,0].plot(ts, ap1.xps)
axs[0,0].set_xlabel('time (s)')
axs[0,0].set_ylabel('X position (m)')
axs[0,0].set_title('X postion and time')
axs[0,0].legend(['Particle 1', 'Particle 2'])

axs[1,0].plot(ts, ap1.yps)
axs[1,0].set_xlabel('time (s)')
axs[1,0].set_ylabel('Y position (m)')
axs[1,0].set_title('Y postion and time')
axs[1,0].legend(['Particle 1', 'Particle 2'])

axs[0,1].plot(ts, ap1.xvs)
axs[0,1].set_xlabel('time (s)')
axs[0,1].set_ylabel('X velocity (m/s)')
axs[0,1].set_title('X velocity and time')
axs[0,1].legend(['Particle 1', 'Particle 2'])

axs[1,1].plot(ts, ap1.yvs)
axs[1,1].set_xlabel('time (s)')
axs[1,1].set_ylabel('Y velocity (m/s)')
axs[1,1].set_title('Y velocity and time')
axs[1,1].legend(['Particle 1', 'Particle 2'])

axs[0,2].plot(ts, ap1.xas)
axs[0,2].set_xlabel('time (s)')
axs[0,2].set_ylabel('X acceleration (m//$s^2$)')
axs[0,2].set_title('X acceleration and time')
axs[0,2].legend(['Particle 1', 'Particle 2'])

axs[1,2].plot(ts, ap1.yas)
axs[1,2].set_xlabel('time (s)')
axs[1,2].set_ylabel('Y acceleration (m//$s^2$)')
axs[1,2].set_title('Y acceleration and time')
axs[1,2].legend(['Particle 1', 'Particle 2'])

fig.tight_layout()
plt.show()
'''

iteration 3.py
import matplotlib.pyplot as plt
import numpy as np

# moving particle and resting particle
class par: # particle class

  #vacuum electric permittivity
  vep = 8.8541878128 * 10**-12

# self, mass, electric charge, x acceleration, y acceleration, x velocity, y velocity, x position, y position
  def __init__(self, m, e, xa, ya, xv, yv, xp, yp): 
    self.m = m # (kg)
    self.e = e # (coulomb)
    self.xa = xa # (m/s^2)
    self.xas = [xa]
    self.ya = ya # (m/s^2)
    self.yas = [ya]
    self.xv = xv # (m/s)
    self.xvs = [xv]
    self.yv = yv # (m/s)
    self.yvs = [yv]
    self.xp = xp # (m)
    self.xps = [xp]
    self.yp = yp # (m)
    self.yps = [yp]
  
  def next_position(self):
    self.xp = self.xp + self.xv * time_interval
    self.xps.append(self.xp)
    self.yp = self.yp + self.yv * time_interval
    self.yps.append(self.yp)

  def next_velocity(self):
    self.xv = self.xv + self.xa * time_interval
    self.xvs.append(self.xv)
    self.yv = self.yv + self.ya * time_interval
    self.yvs.append(self.yv)
  
  def next_acceleration(self, other):
    dx = abs(self.xp - other.xp)
    dy = abs(self.yp - other.yp)
    self.xa = (dx * self.e * other.e) / ((dx**2 + dy**2)**1.5 * 4 * np.pi * par.vep * self.m)
    if self.xp < other.xp:
      self.xa = -self.xa
    self.xas.append(self.xa)
    self.ya = (dy * self.e * other.e) / ((dx**2 + dy**2)**1.5 * 4 * np.pi * par.vep * self.m)
    if self.yp < other.yp:
      self.ya = -self.ya
    self.yas.append(self.ya)

# main computational loop
time_interval = 0.2
total_time = 20
ts = np.arange(0, total_time+time_interval, time_interval)

# alpha particle
apm = 6.6446573357 * 10**-27
apc = 2 * 1.602176634 * 10**-19
hmax = 10
hs = np.arange(0, hmax+0.1, 0.1)
aps = []
for h in hs:
  aps.append(par(apm, apc, 0, 0, 2, 0, 0, h))
  
ap1 = par(apm, apc, 0, 0, 2, 0, 0, 0)
# par2 is a gold nucleus
gnm = 196.966543 * 1.660539 * 10**-27
gnc = 79 * 1.602176634 * 10**-19
gn = par(gnm, gnc, 0, 0, 0, 0, 10, 0)

# main computation loop
for t in ts[1:]:
  i = 0
  for h in hs:
    aps[i].next_position()
    aps[i].next_velocity()
    aps[i].next_acceleration(gn)
    i = i + 1

# plotting the path of all particles
i = 0
for h in hs:
    plt.plot(aps[i].xps, aps[i].yps)
    i = i + 1
plt.plot(10,0,'ro') 
plt.xlabel('X')
plt.ylabel('Y')
plt.title('The path of both particles in 2D')
plt.show()

iteration 4.py
import matplotlib.pyplot as plt
import numpy as np

# moving particle and resting particle
class par: # particle class

  #vacuum electric permittivity
  vep = 8.8541878128 * 10**-12

# self, mass, electric charge, x acceleration, y acceleration, x velocity, y velocity, x position, y position
  def __init__(self, m, e, xa, ya, xv, yv, xp, yp): 
    self.m = m # (kg)
    self.e = e # (coulomb)
    self.xa = xa # (m/s^2)
    self.xas = [xa]
    self.ya = ya # (m/s^2)
    self.yas = [ya]
    self.xv = xv # (m/s)
    self.xvs = [xv]
    self.yv = yv # (m/s)
    self.yvs = [yv]
    self.xp = xp # (m)
    self.xps = [xp]
    self.yp = yp # (m)
    self.yps = [yp]
  
  def next_position(self):
    self.xp = self.xp + self.xv * time_interval
    self.xps.append(self.xp)
    self.yp = self.yp + self.yv * time_interval
    self.yps.append(self.yp)

  def next_velocity(self):
    self.xv = self.xv + self.xa * time_interval
    self.xvs.append(self.xv)
    self.yv = self.yv + self.ya * time_interval
    self.yvs.append(self.yv)
  
  def next_acceleration(self, other):
    dx = abs(self.xp - other.xp)
    dy = abs(self.yp - other.yp)
    self.xa = (dx * self.e * other.e) / ((dx**2 + dy**2)**1.5 * 4 * np.pi * par.vep * self.m)
    if self.xp < other.xp:
      self.xa = -self.xa
    self.xas.append(self.xa)
    self.ya = (dy * self.e * other.e) / ((dx**2 + dy**2)**1.5 * 4 * np.pi * par.vep * self.m)
    if self.yp < other.yp:
      self.ya = -self.ya
    self.yas.append(self.ya)

# main computational loop
time_interval = 0.2
total_time = 20
ts = np.arange(0, total_time+time_interval, time_interval)

# alpha particle
apm = 6.6446573357 * 10**-27
apc = 2 * 1.602176634 * 10**-19
hmax = 50
hs = np.arange(0, hmax+0.01, 0.01)
aps = []
for h in hs:
  aps.append(par(apm, apc, 0, 0, 2, 0, 0, h))
  
ap1 = par(apm, apc, 0, 0, 2, 0, 0, 0)
# par2 is an gold nucleus
gnm = 196.966543 * 1.660539 * 10**-27
gnc = 79 * 1.602176634 * 10**-19
gn = par(gnm, gnc, 0, 0, 0, 0, 10, 0)

# main computation loop
for t in ts[1:]:
  i = 0
  for h in hs:
    aps[i].next_position()
    aps[i].next_velocity()
    aps[i].next_acceleration(gn)
    i = i + 1

# computation loop for finding the angle
angles = []
i = 0
for h in hs:
    angle = np.degrees(np.arctan(aps[i].yv / aps[i].xv))
    if angle < 0:
      angle = 180 + angle
    elif angle == 0:
      angle = 180
    elif angle > 0:
      angle = angle
    angles.append(angle)
    i = i + 1
print(angles)

# plotting the angle against h
i = 0
plt.plot(hs, angles)
plt.ylabel('Resulting angle (°)')
plt.xlabel('H (m)')
plt.title('The resulting angle against H')
plt.show()


iteration 5.py
import matplotlib.pyplot as plt
import numpy as np


# moving particle and resting particle
class par:  # particle class

    # vacuum electric permittivity
    vep = 8.8541878128 * 10 ** -12

    # self, mass, electric charge, x acceleration, y acceleration, x velocity, y velocity, x position, y position
    def __init__(self, m, e, xa, ya, xv, yv, xp, yp):
        self.m = m  # (kg)
        self.e = e  # (coulomb)
        self.xa = xa  # (m/s^2)
        self.ya = ya  # (m/s^2)
        self.xv = xv  # (m/s)
        self.yv = yv  # (m/s)
        self.xp = xp  # (m)
        self.yp = yp  # (m)

    def next_position(self):
        self.xp = self.xp + self.xv * time_interval
        self.yp = self.yp + self.yv * time_interval

    def next_velocity(self):
        self.xv = self.xv + self.xa * time_interval
        self.yv = self.yv + self.ya * time_interval

    def next_acceleration(self, other):
        dx = abs(self.xp - other.xp)
        dy = abs(self.yp - other.yp)
        self.xa = (dx * self.e * other.e) / ((dx ** 2 + dy ** 2) ** 1.5 * 4 * np.pi * par.vep * self.m)
        if self.xp < other.xp:
            self.xa = -self.xa
        self.ya = (dy * self.e * other.e) / ((dx ** 2 + dy ** 2) ** 1.5 * 4 * np.pi * par.vep * self.m)
        if self.yp < other.yp:
            self.ya = -self.ya


# main computational loop
time_interval = 0.2
total_time = 15
ts = np.arange(0, total_time + time_interval, time_interval)

# alpha particle
apm = 6.6446573357 * 10 ** -27
apc = 2 * 1.602176634 * 10 ** -19
hmax = 10000
hs = np.arange(0, hmax + 0.01, 0.01)
aps = []
for h in hs:
    aps.append(par(apm, apc, 0, 0, 2, 0, 0, h))

# atom nucleus
anm = 196.966543 * 1.660539 * 10 ** -27
anc = 79 * 1.602176634 * 10 ** -19
ans = []
cs = []
for cm in range(1, 101):
    ans.append(par(anm, cm * anc, 0, 0, 0, 0, 10, 0))
    cs.append(cm * anc)

# main computation loop

eas = []

for c in range(len(cs)):
    i = 0
    for h in hs:
        for t in ts[1:]:
            aps[i].next_position()
            aps[i].next_velocity()
            aps[i].next_acceleration(ans[c])
        print(h, "           ", np.arctan(aps[i].yv / aps[i].xv))
        if 0.01745329 > np.arctan(aps[i].yv / aps[i].xv) > 0:  # 0.01745329 radians = 1 degree
            eas.append(h)
            break
        i = i + 1

print(
    eas)  # [46.64, 67.03, 82.54, 95.58, 107.04, 117.39, 126.9, 135.75, 144.05, 151.9, 159.37, 166.5, 173.34, 179.92000000000002, 186.26, 192.4, 198.35, 204.13, 209.74, 215.21, 220.55, 225.75, 230.84, 235.83, 240.70000000000002, 245.48000000000002, 250.17000000000002, 254.78, 259.3, 263.74, 268.11, 272.42, 276.65000000000003, 280.82, 284.93, 288.98, 292.97, 296.92, 300.8, 304.64, 308.44, 312.18, 315.88, 319.54, 323.16, 326.74, 330.28000000000003, 333.78000000000003, 337.24, 340.67, 344.07, 347.43, 350.76, 354.06, 357.32, 360.56, 363.77, 366.95, 370.11, 373.24, 376.34000000000003, 379.41, 382.47, 385.49, 388.5, 391.48, 394.44, 397.37, 400.29, 403.18, 406.06, 408.91, 411.74, 414.55, 417.35, 420.13, 422.88, 425.62, 428.35, 431.05, 433.74, 436.41, 439.07, 441.71000000000004, 444.33, 446.94, 449.53000000000003, 452.11, 454.68, 457.23, 459.76, 462.28000000000003, 464.79, 467.29, 469.77, 472.24, 474.69, 477.13, 479.57, 481.98]

# plotting the angle against h
i = 0
plt.plot(eas, cs)
plt.xlabel('H (m)')
plt.ylabel('Charge (coulomb)')
plt.title('The nucleus charge against H')
plt.show()


iteration 6.py
import matplotlib.pyplot as plt
import numpy as np
  
# atom nucleus charge
anc = 79 * 1.602176634 * 10**-19
cs = []
for cm in range(1,101):
  cs.append(cm * anc)

# plotting the angle against h
# eas list was data extracted from iteration 5
eas = [46.64, 67.03, 82.54, 95.58, 107.04, 117.39, 126.9, 135.75, 144.05, 151.9, 159.37, 166.5, 173.34, 179.92000000000002, 186.26, 192.4, 198.35, 204.13, 209.74, 215.21, 220.55, 225.75, 230.84, 235.83, 240.70000000000002, 245.48000000000002, 250.17000000000002, 254.78, 259.3, 263.74, 268.11, 272.42, 276.65000000000003, 280.82, 284.93, 288.98, 292.97, 296.92, 300.8, 304.64, 308.44, 312.18, 315.88, 319.54, 323.16, 326.74, 330.28000000000003, 333.78000000000003, 337.24, 340.67, 344.07, 347.43, 350.76, 354.06, 357.32, 360.56, 363.77, 366.95, 370.11, 373.24, 376.34000000000003, 379.41, 382.47, 385.49, 388.5, 391.48, 394.44, 397.37, 400.29, 403.18, 406.06, 408.91, 411.74, 414.55, 417.35, 420.13, 422.88, 425.62, 428.35, 431.05, 433.74, 436.41, 439.07, 441.71000000000004, 444.33, 446.94, 449.53000000000003, 452.11, 454.68, 457.23, 459.76, 462.28000000000003, 464.79, 467.29, 469.77, 472.24, 474.69, 477.13, 479.57, 481.98]

plt.xscale('log')
plt.yscale('log')
plt.plot(eas,cs)
plt.xlabel('H (m)')
plt.ylabel('Charge (coulomb)')
plt.title('The nucleus charge against H')
plt.show()


iteration 1.py
import matplotlib.pyplot as plt
import numpy as np

# moving particle and resting particle
class par: # particle class

  #vacuum electric permittivity
  vep = 8.8541878128 * 10**-12

# self, mass, electric charge, x acceleration, y acceleration, x velocity, y velocity, x position, y position
  def __init__(self, m, e, xa, ya, xv, yv, xp, yp): 
    self.m = m # (kg)
    self.e = e # (coulomb)
    self.xa = xa # (m/s^2)
    self.xas = [xa]
    self.ya = ya # (m/s^2)
    self.yas = [ya]
    self.xv = xv # (m/s)
    self.xvs = [xv]
    self.yv = yv # (m/s)
    self.yvs = [yv]
    self.xp = xp # (m)
    self.xps = [xp]
    self.yp = yp # (m)
    self.yps = [yp]
  
  def next_position(self):
    self.xp = self.xp + self.xv * time_interval
    self.xps.append(self.xp)
    self.yp = self.yp + self.yv * time_interval
    self.yps.append(self.yp)

  def next_velocity(self):
    self.xv = self.xv + self.xa * time_interval
    self.xvs.append(self.xv)
    self.yv = self.yv + self.ya * time_interval
    self.yvs.append(self.yv)
  
  def next_acceleration(self, other):
    dx = abs(self.xp - other.xp)
    dy = abs(self.yp - other.yp)
    self.xa = (dx * self.e * other.e) / ((dx**2 + dy**2)**1.5 * 4 * np.pi * par.vep * self.m)
    if self.xp < other.xp:
      self.xa = -self.xa
    self.xas.append(self.xa)
    self.ya = (dy * self.e * other.e) / ((dx**2 + dy**2)**1.5 * 4 * np.pi * par.vep * self.m)
    if self.yp < other.yp:
      self.ya = -self.ya
    self.yas.append(self.ya)

# main computational loop
time_interval = 0.2
total_time = 12
ts = np.arange(0, total_time+time_interval, time_interval)

par1 = par(1, 0.00001, 0, 0, 2, 0, 0, 0)
par2 = par(1, 0.00001, 0, 0, 0, -0.1, 10, 1)

for t in ts[1:]:
  par1.next_position()
  par2.next_position()
  par1.next_velocity()
  par2.next_velocity()
  par1.next_acceleration(par2)
  par2.next_acceleration(par1)





'''
# plotting the path of bother particles
plt.plot(par1.xps, par1.yps)
plt.plot(par2.xps, par2.yps)
plt.xlabel('X')
plt.ylabel('Y')
plt.title('The path of both particles in 2D')
plt.show()
'''

# plotting 2d
fig, axs = plt.subplots(2,3)

axs[0,0].plot(ts, par1.xps)
axs[0,0].plot(ts, par2.xps)
axs[0,0].set_xlabel('time (s)')
axs[0,0].set_ylabel('X position (m)')
axs[0,0].set_title('X postion and time')
axs[0,0].legend(['Particle 1', 'Particle 2'])

axs[1,0].plot(ts, par1.yps)
axs[1,0].plot(ts, par2.yps)
axs[1,0].set_xlabel('time (s)')
axs[1,0].set_ylabel('Y position (m)')
axs[1,0].set_title('Y postion and time')
axs[1,0].legend(['Particle 1', 'Particle 2'])

axs[0,1].plot(ts, par1.xvs)
axs[0,1].plot(ts, par2.xvs)
axs[0,1].set_xlabel('time (s)')
axs[0,1].set_ylabel('X velocity (m/s)')
axs[0,1].set_title('X velocity and time')
axs[0,1].legend(['Particle 1', 'Particle 2'])

axs[1,1].plot(ts, par1.yvs)
axs[1,1].plot(ts, par2.yvs)
axs[1,1].set_xlabel('time (s)')
axs[1,1].set_ylabel('Y velocity (m/s)')
axs[1,1].set_title('Y velocity and time')
axs[1,1].legend(['Particle 1', 'Particle 2'])

axs[0,2].plot(ts, par1.xas)
axs[0,2].plot(ts, par2.xas)
axs[0,2].set_xlabel('time (s)')
axs[0,2].set_ylabel('X acceleration (m//$s^2$)')
axs[0,2].set_title('X acceleration and time')
axs[0,2].legend(['Particle 1', 'Particle 2'])

axs[1,2].plot(ts, par1.yas)
axs[1,2].plot(ts, par2.yas)
axs[1,2].set_xlabel('time (s)')
axs[1,2].set_ylabel('Y acceleration (m//$s^2$)')
axs[1,2].set_title('Y acceleration and time')
axs[1,2].legend(['Particle 1', 'Particle 2'])

fig.tight_layout()
plt.show()
